pkg/operator/etcdcertsigner/etcdcertsignercontroller_test.go

```diff
@@ -0,0 +1,168 @@
+package etcdcertsigner
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"github.com/openshift/library-go/pkg/crypto"
+	"math/big"
+	"testing"
+	"time"
+)
+
+func Test_verifyCertificateSAN(t *testing.T) {
+	type certConfig struct {
+		certName    string
+		configHosts []string
+	}
+	tests := []struct {
+		name       string
+		args       certConfig
+		checkHosts []string
+		wantErr    bool
+	}{
+		{
+			name: "Match localhost",
+			args: certConfig{
+				certName:    "PeerCertificate",
+				configHosts: []string{"localhost"},
+			},
+			checkHosts: []string{"localhost"},
+			wantErr:    false,
+		},
+		{
+
+			name: "Match ipaddress in peer certs",
+			args: certConfig{
+				certName: "PeerCertificate with IP",
+				configHosts: []string{
+					"localhost",
+					"1.2.3.4",
+				},
+			},
+			checkHosts: []string{"1.2.3.4"},
+			wantErr:    false,
+		},
+		{
+
+			name: "Mismatch ipaddress in peer certs",
+			args: certConfig{
+				certName: "PeerCertificate with IP",
+				configHosts: []string{
+					"localhost",
+					"1.2.3.4",
+				},
+			},
+			checkHosts: []string{"2.3.4.1"},
+			wantErr:    true,
+		},
+		{
+			name: "Match ipaddress in server certs",
+			args: certConfig{
+				certName: "ServerCertificate with IP",
+				configHosts: []string{
+					"localhost",
+					"etcd.kube-system.svc",
+					"etcd.kube-system.svc.cluster.local",
+					"etcd.openshift-etcd.svc",
+					"etcd.openshift-etcd.svc.cluster.local",
+					"127.0.0.1",
+					"::1",
+					"0:0:0:0:0:0:0:1",
+					"localhost",
+					"1.2.3.4",
+				},
+			},
+			checkHosts: []string{"1.2.3.4"},
+			wantErr:    false,
+		},
+		{
+			name: "Match IPv6 addr in server certs",
+			args: certConfig{
+				certName: "ServerCert with IPv6",
+				configHosts: []string{
+					"localhost",
+					"etcd.kube-system.svc",
+					"etcd.kube-system.svc.cluster.local",
+					"etcd.openshift-etcd.svc",
+					"etcd.openshift-etcd.svc.cluster.local",
+					"127.0.0.1",
+					"::1",
+					"0:0:0:0:0:0:0:1",
+					"localhost",
+					"2001:0DB8:C21A",
+				},
+			},
+			checkHosts: []string{"2001:0DB8:C21A"},
+			wantErr:    false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			certData := makeCerts(t, tt.args.configHosts)
+			if err := verifyCertificateSAN(certData, tt.args.certName, tt.checkHosts); (err != nil) != tt.wantErr {
+				t.Errorf("verifyCertificateSAN() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func makeCerts(t *testing.T, hosts []string) map[string][]byte {
+	// below code is copied from vendor/github.com/openshift/library-go/pkg/crypto/crypto.go
```

Comment:
I'm assuming the code was copied because library-go didn't expose the functionality for reuse. Could library-go be updated to allow reuse?

Reply:
Yeah. I agree.